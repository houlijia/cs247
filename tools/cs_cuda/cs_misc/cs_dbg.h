#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
void dbg_pdata_i( char *s, int *dp, int size ) ;
void dbg_pdata_c( char *s, char *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( char *s, int *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
void dbg_p_d_data_i_mn_skip ( char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_i_mn_v2 ( char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

// tsting ...
void dbg_p_d_data_ll ( char *s, long long *dp, int size ) ;

#endif 
