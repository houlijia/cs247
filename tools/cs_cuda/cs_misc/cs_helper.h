#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void clear_device_mem_i ( int *d_a, int n ) ;
void set_device_mem_i ( int *d_a, int n, int val ) ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;
void h_expand_c_to_i ( char *din, int *dout, int n ) ;	

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_i ( int *dp, int *hp, int size ) ;

#endif 
