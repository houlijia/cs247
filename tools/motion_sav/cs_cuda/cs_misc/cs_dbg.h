#ifndef __CS_DBG_H__ 
#define __CS_DBG_H__ 

#include "cs_header.h"

// #define CUDA_DBG 

int dbg_init( int ) ;
void dbg_clear_buf( int *cp, int size ) ;
void dbg_set_buf( int *cp, int size, int set ) ;
int dbg_put_d_data ( char *dp, char *hp, int size )  ;
int dbg_get_d_data ( char *dp, char *hp, int size )  ;
int dbg_copy_d_data ( char *dtp, char *dfp, int size ) ;
void dbg_pdata_i( const char *s, const int *dp, int size ) ;
void dbg_pdata_f( const char *s, const float *dp, int size ) ;
void dbg_pdata_c( const char *s, const char *dp, int size ) ;
void dbg_pdata_d( const char *s, const float *dp, int size ) ;
void dbg_mdata( int *dp, int size ) ;
void dbg_p_d_data_c ( const char *s, char *dp, int size ) ;
void dbg_p_d_data_i ( const char *s, int *dp, int size ) ;
void dbg_p_d_data_f ( const char *s, float *dp, int size ) ;
void dbg_p_d_data_d ( const char *s, float *dp, int size ) ;
char *dbg_d_malloc_c ( int size ) ;
int *dbg_d_malloc_i ( int size ) ;
float *dbg_d_malloc_f ( int size ) ;

void dbg_p_d_data_f_mn_skip ( const char *s, float *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;
void dbg_p_d_data_i_mn_skip ( const char *s, int *dp, int size, int m, int n, int z,
	int doprint, int perm_size ) ;

void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_f_mn ( const char *s, float *dp, int size, int m, int n, int doprint, int printrow ) ;
void dbg_p_d_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;
void dbg_p_d_data_c_mn ( const char *s, char *dp, int size, int m, int n, int doprint ) ;
void dbg_p_data_i_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

// print the motion detection T/V/H/Va format
void dbg_p_data_md_f_mn ( const char *s, int *dp, int size, int m, int n, int doprint ) ;

void dbg_p_d_data_f_mn_v2 ( const char *s, float *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_d_data_i_mn_v2 ( const char *s, int *devp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;
void dbg_p_data_i_mn_v2 ( const char *s, int *hp, int size, int doprint,
	struct cube *dp, int blk_in_x, int blk_in_y ) ;

void dbg_p_d_data_i_cube ( const char *s, int *devp, int vx, int hy, int tz ) ;
void dbg_p_d_data_i_cube ( const char *s, float *devp, int vx, int hy, int tz ) ;
void dbg_p_data_i_cube ( const char *s, float *dp, int vx, int hy, int tz ) ;

void dbg_pr_first_last ( char *s, float *d_p, int len, int pr_size ) ;
void dbg_pr_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, int *d_p, int len, int pr_size ) ;
void dbg_pr_h_first_last ( char *s, char *h_p, int len, int pr_size ) ;

// tsting ...
void dbg_p_d_data_ll ( const char *s, long long *dp, int size ) ;

// for random
int dbg_ck_unique ( char *s, int *dp, int size ) ;
int dbg_perm_ck ( int callid, int callid2, int zero_in_zero, int *d_bp, int size ) ;

// TVH
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print, int do_skip ) ;
void cs_p_d_tvh ( const char *s, int *dmem, int record_size, int num_rec, int do_print,
	int do_skip, int start ) ;

#endif 
