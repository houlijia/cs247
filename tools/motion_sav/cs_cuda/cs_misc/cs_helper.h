#ifndef __CS_HELPER_H__
#define __CS_HELPER_H__

int max_log2( int ) ;

// clear the device mem ... n in size of int
void set_device_mem_i ( int *d_a, int n, int val ) ;
void clear_device_mem_i ( int *d_a, int n ) ;

void set_device_mem ( float *d_a, int n, float val ) ;
void clear_device_mem ( float *d_a, int n )	 ;

// clear the device mem ... n in size of char
void clear_device_mem_c ( char *d_a, int n ) ;
void set_device_mem_c ( char *d_a, int n, char c ) ;

void h_expand_c_to_i ( char *din, int *dout, int n ) ;	
void h_expand_c_to_i ( unsigned char *din, int *dout, int n ) ;	
template<typename F, typename T> void h_cast_T_to_T ( F *din, T *dout, int n ) ;

// block adjustment
void h_block_adj( int n, int nThreadsPerBlock, int *block );
//

// omp timer 
int omp_timer_init( int ) ;
int omp_timer_on( int ) ;
int omp_timer_off( int ) ;
int omp_timer_get( int, double *dp, int *cp, double *ddp ) ;

// opt
int get_nums( int ac, char *av[], int idx, int cnt, int *np ) ;
unsigned int htoi(const char s[]) ;

// cs timer 
int cs_timer_init( int cnt ) ;
int cs_timer_on( int idx ) ;
int cs_timer_off( int idx ) ;
int cs_timer_get( int idx, clock_t *sp, clock_t *up, int *cp,
	double *asp, double *aup ) ;

// swap
void htonl_device_mem_i ( int *d_a, int n )	;

// tsting ...
void h_tst_longlong ( long long *d_a, int n )	 ;

// abs a vector

int h_set_abs ( int *d_a, int n, int record_len, int skip )	 ;

// set config
void h_set_config( struct cs_xyz *d_a, struct cube *hp ) ;
int h_set_cube_config( struct cube *d_a, struct cube *hp ) ;
void h_p_d_cube_config ( const char *, struct cube *dp ) ;
void h_p_h_cube_config ( const char *, struct cube *dp ) ;

// weight shift
int weight_sft( int scheme, int size, int m, int n ) ;

// generic read/write int from/to device ... for ret code
int put_d_data_i ( int *dp, int *hp, int size ) ;
int put_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes
int get_d_data_i ( int *dp, int *hp, int size ) ;
int get_d_data_f ( float *dp, float *hp, int size ) ;	// size in bytes

void h_do_int_to_float ( int *intp, float *fp, int tbl_size ) ;

#endif 
