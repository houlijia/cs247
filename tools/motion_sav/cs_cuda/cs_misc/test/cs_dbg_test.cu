
#include <iostream>
using namespace std;

#include <thrust/reduce.h>
#include <thrust/sequence.h>
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <math.h>

#include "cs_dbg.h"
#include "cs_helper.h"

#define CUDA_DBG
#define BUF_SIZE_INT		30

float *fp1 ;

float dbuf[] = { 0.1, 2.3, 3.0, 4.7 } ;
float dbuf1[ BUF_SIZE_INT ] ;

int *ip, ibuf[ BUF_SIZE_INT ] ;
char *cp, cbuf[ BUF_SIZE_INT ] ;

int
main( int ac, char *av[] )
{
	int i, k ;

	setbuf( stdout, NULL ) ;
	setbuf( stderr, NULL ) ;

	dbg_init ( 1024 * 1024 ) ;

	if (( k = cudaMalloc( &fp1, sizeof ( float ) * BUF_SIZE_INT )) != cudaSuccess )
	{
		printf("%s: cube alloc failed %d \n", __func__, k ) ;
		exit ( 0 ) ;
	}

	put_d_data_f ( fp1, dbuf, sizeof ( float ) *4 ) ; 
	dbg_p_d_data_f ( "print 4 ", fp1, 4 ); 

	get_d_data_f ( fp1, dbuf1, 4 * sizeof ( float )) ;
	dbg_pdata_f ( "print 4 ", dbuf1, 4 ); 

	for ( i = 0 ; i < BUF_SIZE_INT ; i++ )
		ibuf[i] = i ;

	ip = ( int * ) fp1; 

	put_d_data_i ( ip, ibuf, BUF_SIZE_INT * sizeof ( int )) ;

	dbg_pr_first_last( "test first_last", ip, BUF_SIZE_INT, 10 ) ;

	dbg_pr_h_first_last( "test first_last", ibuf, BUF_SIZE_INT, 10 ) ;

	for ( i = 0 ; i < BUF_SIZE_INT ; i++ )
		cbuf[i] = i ;

	cp = ( char * ) fp1 ;

	put_d_data_i ( ( int * )cp, ( int  *)cbuf, BUF_SIZE_INT * sizeof ( char )) ;
	dbg_pr_h_first_last( "test first_last", cbuf, BUF_SIZE_INT, 10 ) ;
}
