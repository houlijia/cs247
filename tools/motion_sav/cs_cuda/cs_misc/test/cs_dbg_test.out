Wed Feb 25 14:48:18 EST 2015
cs_dbg_test
---------------------------------------------------
dbg_p_d_data_f: print 4  size 16 dp 0x1300200000
dbg_pdata_f: print 4 
0 -- 0.100000
1 -- 2.300000
2 -- 3.000000
3 -- 4.700000
dbg_pdata_f: print 4 
0 -- 0.100000
1 -- 2.300000
2 -- 3.000000
3 -- 4.700000
dbg_p_d_data_i: test first_last : FIRST 10 OF 30 size 40 dp 0x1300200000
dbg_pdata_i: test first_last : FIRST 10 OF 30
0 -- 00000000 0
1 -- 00000001 1
2 -- 00000002 2
3 -- 00000003 3
4 -- 00000004 4
5 -- 00000005 5
6 -- 00000006 6
7 -- 00000007 7
8 -- 00000008 8
9 -- 00000009 9
dbg_p_d_data_i: test first_last : CENTER 10 AFTER 15 size 40 dp 0x130020003c
dbg_pdata_i: test first_last : CENTER 10 AFTER 15
0 -- 0000000f 15
1 -- 00000010 16
2 -- 00000011 17
3 -- 00000012 18
4 -- 00000013 19
5 -- 00000014 20
6 -- 00000015 21
7 -- 00000016 22
8 -- 00000017 23
9 -- 00000018 24
dbg_p_d_data_i: test first_last : LAST 10 AFTER 20 size 40 dp 0x1300200050
dbg_pdata_i: test first_last : LAST 10 AFTER 20
0 -- 00000014 20
1 -- 00000015 21
2 -- 00000016 22
3 -- 00000017 23
4 -- 00000018 24
5 -- 00000019 25
6 -- 0000001a 26
7 -- 0000001b 27
8 -- 0000001c 28
9 -- 0000001d 29
dbg_pdata_i: test first_last : FIRST 10 OF 30
0 -- 00000000 0
1 -- 00000001 1
2 -- 00000002 2
3 -- 00000003 3
4 -- 00000004 4
5 -- 00000005 5
6 -- 00000006 6
7 -- 00000007 7
8 -- 00000008 8
9 -- 00000009 9
dbg_pdata_i: test first_last : CENTER 10 AFTER 15
0 -- 0000000f 15
1 -- 00000010 16
2 -- 00000011 17
3 -- 00000012 18
4 -- 00000013 19
5 -- 00000014 20
6 -- 00000015 21
7 -- 00000016 22
8 -- 00000017 23
9 -- 00000018 24
dbg_pdata_i: test first_last : LAST 10 AFTER 20
0 -- 00000014 20
1 -- 00000015 21
2 -- 00000016 22
3 -- 00000017 23
4 -- 00000018 24
5 -- 00000019 25
6 -- 0000001a 26
7 -- 0000001b 27
8 -- 0000001c 28
9 -- 0000001d 29
dbg_pdata_c: test first_last : FIRST 10 OF 30
0 -- 00 0
1 -- 01 1
2 -- 02 2
3 -- 03 3
4 -- 04 4
5 -- 05 5
6 -- 06 6
7 -- 07 7
8 -- 08 8
9 -- 09 9
dbg_pdata_c: test first_last : CENTER 10 AFTER 15
0 -- 0f 15
1 -- 10 16
2 -- 11 17
3 -- 12 18
4 -- 13 19
5 -- 14 20
6 -- 15 21
7 -- 16 22
8 -- 17 23
9 -- 18 24
dbg_pdata_c: test first_last : LAST 10 AFTER 20
0 -- 14 20
1 -- 15 21
2 -- 16 22
3 -- 17 23
4 -- 18 24
5 -- 19 25
6 -- 1a 26
7 -- 1b 27
8 -- 1c 28
9 -- 1d 29
